Unstar all messages in topic.

abhijeetbodas2001 opened the pull request:
The previous attempt at this #16898 was buggy and hence was reverted in #17429.
This PR is another attempt which addresses the bugs noticed there.
This also adds node tests for `static/starred_messages.js`.


https://user-images.githubusercontent.com/55339528/110300208-a0ca8a80-801c-11eb-98f5-5b972565a4c6.mp4

abhijeetbodas2001 commented:
Resolved conflicts in the last push.

timabbott commented on static/js/message_flags.js:
```diff
@@ -4,12 +4,12 @@ import * as channel from "./channel";
 import * as starred_messages from "./starred_messages";
 
-function send_flag_update(message, flag, op) {
+function update_flag_for_msgs(msg_ids, flag, op) {
```
I think it'd be helpful for this to still contain `send` in its name, as I recall in the past some confusion between this function and others that update the UI (after receiving flag updates from elsewhere).

abhijeetbodas2001 replied:
Renamed this to `send_flag_update_for_msgs`.

timabbott commented on frontend_tests/node_tests/starred_messages.js:
```diff
@@ -38,6 +40,48 @@ run_test("remove starred", (override) => {
+    assert.deepEqual(starred_messages.get_starred_msg_ids_in_topic(undefined, "topic name"), []);
+    assert.deepEqual(starred_messages.get_starred_msg_ids_in_topic(3, undefined), []);
+
+    override(message_store, "get", (id) => {
```
Can you add a brief comment explaining the design of this setup?  It's complex enough that I think it'd be useful for understanding the test, both for review now and folks working on the test in the future.

timabbott replied:
(Also, would it be simpler to just insert messages 2-5, maybe with a few more fields added, into `message_store`, rather than mocking the `get` function?  Generally we prefer that sort of test where convenient).

If that doesn't feel easy, we may want to add a common test helper `create_mock_message()` that takes just id/type/recipient as parameters and fills out plausible values for all the other fields to make a real message object; I'm sure we'd find many other places to use such a helper.

abhijeetbodas2001 replied:
I added the `create_mock_message` helper.

> and fills out plausible values for all the other fields

I left this part out, the reasoning being that it would reduce the utility of this helper, and I think it would be better for test readability to just add all fields required in the test itself. I've written more about this in the commit message, but please correct me if I'm wrong! In this commit I also changed some existing instances of `messa_store.get()` mocks to use this new helper method. One more thing -- I added this helper in `static/message_store.js` itself, should it be somewhere else, or this is a fine place for it? (I added a comment in there saying that its for use in tests only, though)

timabbott commented on static/js/stream_popover.js:
```diff
@@ -431,6 +435,32 @@ export function register_stream_handlers() {
+
+    $("body").on("click", "#do_unstar_messages_in_topic_button", (e) => {
+        const topic_name = $("#unstar-messages-modal").attr("data-topic-name");
+        const stream_id = $("#unstar-messages-modal").attr("data-stream-id");
```
I wonder if it'd be better to store these values in a lambda/continuation, rather than in the DOM this way -- I guess that'd be easy to od if the `unstar-messages-modal` wasn't a singleton (and we could just register the click handler inside the function that pops it open).

abhijeetbodas2001 replied:
Both the unstar all messages and unstar all messages in topic systems now use the `confirm_dialog.js` widget! This took a lot of code out of `stream_popover.js`, which anyways wasn't related to the stream popovers directly.

abhijeetbodas2001 commented:
@timabbott Thanks for the review! I updated the PR with the following changes-
1. Added 3 new commits
    - 673e22b23e2bf38d2f38cba4e60a24df9a04d690 addresses a small bug/discrepancy I noticed while working on this. This is at the beginning of the PR.
    - b37d3f785ebc686f2d0d58f9132b83af7463834e adds the helper as you suggested in the [comment above](https://github.com/zulip/zulip/pull/17440#discussion_r587925685). 
       > and fills out plausible values for all the other fields

       I left this part out, the reasoning being that it would restrict the use of this helper, and I think it would be better for test readability to not add more fields in the helper. I've written more about this in the commit message, but please correct me if I'm wrong! In this commit I also changed some existing instances of `get()` mocks to use this new helper method. One more thing -- I added this helper in `static/message_store.js` itself, should it be somewhere else, or this is a fine place for it? (I added a comment in there saying that its for use in tests only, though)
    - 3fefa525c2337c21e8b22a2f8bf95b180daf3d9c refactors the current unstar-all modal code so that click-handlers for the modal are registered in the same function which opens it.
  
2. Changes in old commits are
    - (Re) renamed `send_flag_update` to `send_flag_update_for_msgs` in d67ab725063cb8da904a2bcc7f43e5450ecbe8df
    - New code for unstar-all-in-topic in b219313ac39bbc7f1620226f6bbcf3184ac0a607 now uses the style described [above](https://github.com/zulip/zulip/pull/17440#discussion_r587926257) (use `on_confirm` function instead of storing the data in DOM). I agree that it is cleaner this way, so that `stream_popovers.js` contains click-handlers only directly involved with the left sidebar, and this also avoids some code-duplication 🙂

abhijeetbodas2001 commented:
@timabbott Thanks for the review! This is ready for another look. Here are the changes since the last review-
1. Added 3 new commits
    - 8c18676508de04fcf6138c52fc2e93b61180498f addresses a small bug/discrepancy I noticed while working on this. This is at the beginning of the PR.
    - c0faed6ba9248255851aedb51a819feb3e872d33 adds the helper as you suggested in the [comment above](https://github.com/zulip/zulip/pull/17440#discussion_r587925685). 
       > and fills out plausible values for all the other fields

       I left this part out, the reasoning being that it would restrict the use of this helper, and I think it would be better for test readability to not add more fields in the helper. I've written more about this in the commit message, but please correct me if I'm wrong! In this commit I also changed some existing instances of `get()` mocks to use this new helper method. One more thing -- I added this helper in `static/message_store.js` itself, should it be somewhere else, or this is a fine place for it? (I added a comment in there saying that its for use in tests only, though)
    - e3844fbd340d9ef111edda22931baec86c3b4dc2 replaces the current unstar-all modal with the `confirm_dialog` Widget..
  
2. Changes in old commits are
    - (Re) renamed `send_flag_update` to `send_flag_update_for_msgs` in ee68a60e7ea4aed687539101f36ccc81cda60626
    - New code for unstar-all-in-topic in a8c84f58e6ac735d0a5dc37ea04ac28141e1c746 also uses the widget from `static/confirm_dialog.js`

abhijeetbodas2001 commented:
I resolved conflicts in the last push. @timabbott this is ready for another review, please take a look!

abhijeetbodas2001 commented:
@timabbott This is ready for another review! I wrote the changes inline, plus added one new commit at the beginning of the PR as a small bugfix.

timabbott commented on static/js/message_flags.js:
```diff
@@ -6,12 +6,12 @@ import * as starred_messages from "./starred_messages";
 import * as unread_ops from "./unread_ops";
 
-function send_flag_update(message, flag, op) {
+function send_flag_update_for_msgs(msg_ids, flag, op) {
```
This should be `.._for_messages` -- we don't use `msgs` in the codebase.

abhijeetbodas2001 replied:
Thanks for pointing out. I'll keep that in mind 🙂

timabbott commented on static/js/message_store.js:
```diff
@@ -32,6 +32,12 @@ export function user_ids() {
+    // For use in tests only. `id` is a required field,
+    // everything else is optional, as required in the test.
+    stored_messages.set(message.id, message);
+}
```
I think this function doesn't fully do what I'd had in mind -- the ideal version would populate values that were not provided in some reasonable way such that we end up with a `message` object with all the normal fields it should have.  But I'm OK with merging this for now since this commit improves things by removing overrides.

@showell FYI in case you decide to do one more thing before taking a break from node tests (or want to help 
@abhijeetbodas2001 do such a project).

timabbott commented on static/templates/unstar_messages_modal.hbs:
```diff
@@ -1,13 +0,0 @@
-        <button type="button" class="button rounded" data-dismiss="modal">{{t "Cancel" }}</button>
-        <button type="button" class="button rounded btn-danger" id="do_unstar_messages_button">{{t "Unstar all messages" }}</button>
-    </div>
-</div>
```
It's nice to delete such a big block of code.  If you're looking for a mini-project, it's worth doing an audit for whether we can do this treatment in a separate PR to clean up other confirm dialogs.

abhijeetbodas2001 replied:
Yep! I'll try to do a follow up PR.

timabbott commented on static/templates/confirm_dialog.hbs:
```diff
@@ -1,4 +1,4 @@
-<div class="modal modal-bg hide fade" id="confirm_dialog_modal" tabindex="-1" role="dialog" aria-labelledby="confirm_dialog_modal" aria-hidden="true">
+<div class="modal new-style modal-bg hide fade" id="confirm_dialog_modal" tabindex="-1" role="dialog" aria-labelledby="confirm_dialog_modal" aria-hidden="true">
```
Why do we need `new-style` added here?  The commit message doesn't comment on this detail.

abhijeetbodas2001 replied:
Ah, my bad, I missed writing about this. I pulled one commit out for this (with the reasoning in the commit message), which also includes another small change which will allow us to use `confirm_dialog.js` more extensively. 

`new-style` has many styles for buttons etc. Here's how the modal looks like without this class...

![image](https://user-images.githubusercontent.com/55339528/111866537-30185c00-8994-11eb-8f45-78d8960f9408.png)
This wasn't a problem earlier, because `confirm_dialog` was used only in the settings pane, which already has this class in the main `settings_overlay.hbs` file.

timabbott commented on static/js/starred_messages.js:
```diff
@@ -43,6 +45,30 @@ export function get_starred_msg_ids() {
+            // but the message itself hasn't been fetched from the server yet.
+            // The message with this `id` may belong to some other topic,
+            // so don't return it.
+            return false;
```
I think we should mention here that this means that our implementation is not correct, and that to fix this we will need to ask the server for the set of message IDs to modify.

(I think it's OK to merge this feature with that known bug and address that as a follow-up, to manage complexity and because I expect this bug to be not easy to trigger).

abhijeetbodas2001 replied:
Updated the comment and commit message to mention this. I think it was decided that the feature would be frontend only in V1? https://chat.zulip.org/#narrow/stream/137-feedback/topic/unstar.20in.20topic/near/735823

timabbott commented:
I merged the first several commits as the series ending with 91234b2a86f026c575f7b5eb60d0df4507b2e4ee, and then commented on the last bit.  Thanks for all your work on this so far @abhijeetbodas2001!

abhijeetbodas2001 commented:
@timabbott Thanks for the review! This is ready for another look. Apart from the inline comments, here are the changes since the last review.
- Added a small bugfix at the beginning of this PR.
- In the video in the PR description comment above, a lag can be seen after clicking on "Unstar messages". I thought earlier that it was my machine being slow, but that isn't the case. I've added a commit (the second one in this PR) which fixes this.
Here's a GIF after this change-
![unstar](https://user-images.githubusercontent.com/55339528/111871013-ff452080-89ad-11eb-99bc-0d8f5fe4b07d.gif)

- Put `starred_messages.js` back in `test-js-with-node/EXEMPT_FILES` -- it isn't a data only module anymore, and I couldn't figure out how to write tests for this. Is that fine, or should I try to write `zjquery` tests?

abhijeetbodas2001 commented:
@timabbott Thanks for the review! This is ready for another look. Apart from the inline comments, here are the changes since the last review.
- Added a small bugfix at the beginning of this PR.
- In the video in the PR description comment above, a lag can be seen after clicking on "Unstar messages". I thought earlier that it was my machine being slow, but that isn't the case. I've added a commit (the second one in this PR) which fixes this.
<details><summary>Here's a GIF after this change</summary>
<img src="https://user-images.githubusercontent.com/55339528/111871013-ff452080-89ad-11eb-99bc-0d8f5fe4b07d.gif" />
</details>

- Put `starred_messages.js` back in `test-js-with-node/EXEMPT_FILES` -- it isn't a data only module anymore, and I couldn't figure out how to write tests for this, nor could I find any examples in the codebase which test other widgets. Is that fine, or should I try to write `zjquery` tests?

timabbott commented:
Merged, thanks @abhijeetbodas2001!  I think this is a reasonable checkpoint even though we'll want to do some follow-up work, which I'd appreciate your driving.
* Can you open an issue for making the `starred_messages.js` logic fetch the set of messages to unstar from the server?  Most simply, this would basically entail doing a `GET /messages` query with `is:starred, stream:, and topic:` filters and throwing away the result aside from the returned message IDs.
* For test coverage, I think the right model would be split out a `starred_messages_ui.js` module that contains the confirm dialog (etc.) -- similar to various commits @showell has been doing this week for other modules (for a different reason).  I think that's probably a reasonable path forward there -- feel free to just submit a PR without an issue since it's pretty small.

abhijeetbodas2001 commented:
I opened #17790 and #17793 respectively for these.
I'm also working on the confirm dialog sweep and will try to open a PR for it soon.
